### DISCUSSION TITLE
Strange behavior

### DISCUSSION THREAD

**Discussion by rkatikam** [highfestiva/finplot#507] [Q&A] [answered]
Hi,
I have a piece of code where I display the candles upto a chosen date in the past e.g 2023-SEP-06 and then as I press the right key / left key it adds/removes 1 bar at a time. A strange behaviour happens  when I move say n times to the right and then move to the left in which case 2 bars are removed instead of 1 bar and if I continue moving left, it keeps removing 1 at a time and if I reverse to the right, it adds 2 bars and then 1 bar for subsequent tight moves. Not sure why this adding/removing of 2 bars happen when the direction changes. To debug I have printed the indices on the console and I see the indices are incrementing/decrementing correctly and hence the data that needs to be displayed is also correct.

My suspicion is around this part of the code 
```
for k in data:
        if data[k] is not None:
            plots[k].update_data(data[k], gfx=False)    
    for k in data:
        if data[k] is not None:
            plots[k].update_gfx()

```
in the function update_plot().

Most of the code is based on the example provided in complicated.py 
I am pasting the full code as well 

`#!/usr/bin/env python3

import finplot as fplt
from math import nan
import pandas as pd
from PyQt6.QtCore import QDateTime, Qt
import pyqtgraph as pg
import yfinance as yf
import datetime as dt
from threading import Lock

stock_ohlcv_data=None
backTestDate_Idx=None
mutex=None

def load_price_history(symbol, interval,backtest_date):
    """Use memoized, and if too old simply load the data."""
    global backTestDate_Idx, stock_ohlcv_data
    
    print("Getting data for: ",symbol)
    print("Back test Date: ",backtest_date)    
    stock_ohlcv_data = yf.download(symbol + ".NS",interval=interval)
    #print(stock_ohlcv_data)
    
    backTestDate_Idx = stock_ohlcv_data.index.get_loc(backtest_date)
    print("Starting Index of the Backtes Date:",backTestDate_Idx)
    stock_ohlcv_data.dropna()
    stock_ohlcv_data["time"] = pd.to_datetime(stock_ohlcv_data.index)
    stock_ohlcv_data = stock_ohlcv_data.drop(["Adj Close"], axis=1)    
    return stock_ohlcv_data
    

def calc_plot_data(df):
    #Returns data for all plots and for the price line.    
    price = df["Open Close High Low".split()]
    volume = df["Open Close Volume".split()]    
    plot_data = dict(
        price=price,
        volume=volume        
    )
    return plot_data

def update_plot():
    global backTestDate_Idx
    
    print("EndingIndex=",backTestDate_Idx)
    # calculate the new plot data    
    data  = calc_plot_data(stock_ohlcv_data[:backTestDate_Idx])
    
    # first update all data, then graphics (for zoom rigidity)    
    for k in data:
        if data[k] is not None:
            plots[k].update_data(data[k], gfx=False)
    
    for k in data:
        if data[k] is not None:
            plots[k].update_gfx()    
    
    

def update_chart():
    
    # remove any previous plots
    ax.reset()
    axo.reset()    
    
    # calculate plot data
    df=stock_ohlcv_data[:backTestDate_Idx]
    data = calc_plot_data(df)
    
    # plot data
    global plots
    plots = {}
    plots["price"] = fplt.candlestick_ochl(data["price"], ax=ax)
    plots["volume"] = fplt.volume_ocv(data["volume"], ax=axo)    
    ax.set_visible(xaxis=True)
    
    # restores saved zoom position, if in range
    fplt.refresh()
    

def on_key(event):
    global backTestDate_Idx,mutex
    keyPressed=event.key()
    mutex.acquire()
    # test for a specific key
    if keyPressed == Qt.Key.Key_Right:        
        backTestDate_Idx+=1        
        print("Right Key. New Index=",backTestDate_Idx)
        update_plot()
        
    elif keyPressed == Qt.Key.Key_Left:        
        backTestDate_Idx-=1        
        print("Left Key. New Index=",backTestDate_Idx)
        update_plot()
    print("update_plot() done")
    mutex.release()
    #update_plot()
   
        
    

if __name__=="__main__":
    plots = {}
    fplt.y_pad = 0.07  # pad some extra (for control panel)
    fplt.max_zoom_points = 7    
    ax= fplt.create_plot("Study", rows=1, init_zoom_periods=300)
    axo = ax.overlay()    
    ax.vb.win.keyPressEvent = on_key
    mutex=Lock()
    load_price_history(symbol="TCS", interval="1D",backtest_date= "2023-09-06")
    update_chart()
    fplt.show()
`

**Comment by highfestiva** [comment, marked as answer]
This was caused by an off-by-one bug, and is fixed in 203d238.

**Reply by rkatikam** [reply]
Thanks

**Comment by rkatikam** [comment]
Again I ran into an issue. If i have multiple panes to display and hence multiple axes, as I press right key, one bar at a time get added. But  the zoom factor changes and hence just after 4 right key presses both the charts are in a zoomed state. I am attaching the full code below.


`
import sys
from PyQt6.QtWidgets import QApplication, QMainWindow, QLabel, QWidget
from PyQt6.QtGui import QKeyEvent
from pyqtgraph.dockarea import DockArea, Dock
from PyQt6.QtCore import QDateTime, Qt
import yfinance as yf
import pandas as pd
import finplot as fplt
stock_ohlcv_data=None
index_ohlcv_data=None
backtestdate_Idx=None
backtestdate_Idx1=None
plots={}
def load_price_history(symbol, interval,backtest_date):   
    global backtestdate_Idx, stock_ohlcv_data,index_ohlcv_data,backtestdate_Idx1
    #print("Getting data for: ",symbol)
    #print("Back test Date: ",backtest_date)    
    stock_ohlcv_data = yf.download(symbol + ".NS",interval=interval)        
    stock_ohlcv_data.dropna()
    stock_ohlcv_data["time"] = pd.to_datetime(stock_ohlcv_data.index)
    stock_ohlcv_data = stock_ohlcv_data.drop(["Adj Close"], axis=1)

    #print(stock_ohlcv_data)    
    backtestdate_Idx = stock_ohlcv_data.index.get_loc(backtest_date)
    #print("Starting Index of the Backtes Date:",backtestdate_Idx)

    index_ohlcv_data = yf.download("^NSEI",interval=interval)
    index_ohlcv_data.dropna()
    index_ohlcv_data["time"] = pd.to_datetime(index_ohlcv_data.index)
    index_ohlcv_data = index_ohlcv_data.drop(["Adj Close"], axis=1)
    #print(index_ohlcv_data)
    backtestdate_Idx1 = index_ohlcv_data.index.get_loc(backtest_date)
    #return stock_ohlcv_data
def compute_data():
    global backtestdate_Idx,backtestdate_Idx1
    
    pricedf=stock_ohlcv_data[:backtestdate_Idx]
    indexdf=index_ohlcv_data[:backtestdate_Idx1]
    #print(pricedf)
    #print(indexdf)
    price = pricedf ["Open Close High Low".split()]
    ma20 = pricedf.Close.rolling(20).mean()
    ma50 = pricedf.Close.rolling(50).mean()
    volume = pricedf ["Open Close Volume".split()]
    index= indexdf["Open Close High Low".split()]
    
    plot_data= {'stockprice':price,
            'ma20':ma20,
            'ma50':ma50,
            'volume':volume,
            'indexprice':index,
            }    
    return plot_data
def on_key(event):
    global backtestdate_Idx,backtestdate_Idx1
    # test for a specific key
    if event.key() == Qt.Key.Key_Right:        
        backtestdate_Idx+=1
        backtestdate_Idx1+=1        
        print("Right Key. New Index=",backtestdate_Idx)
        update_plot()
        #update_chart()
        
    elif event.key() == Qt.Key.Key_Left:        
        backtestdate_Idx-=1
        backtestdate_Idx1-=1        
        print("Left Key. New Index=",backtestdate_Idx)
        update_plot()
        #update_chart()
def update_plot():    
    data=compute_data()
    #print(data)
    for k in data: 
        if data[k] is not None:
            plots[k].update_data(data[k], gfx=False)
    for k in data:
        if data[k] is not None:            
            plots[k].update_gfx()            
    
            
def update_chart():    
    ax0.reset() # remove previous plots    
    ax1.reset() # remove previous plots
    
    data=compute_data()
    global plots
    
    
    plots["ma20"] = fplt.plot(data['ma20'], legend = "MA-20", ax = ax0)
    plots["ma50"] =fplt.plot(data['ma50'], legend = "MA-50", ax = ax0)
    plots["stockprice"]=fplt.candlestick_ochl(data['stockprice'], ax = ax0)
    plots["volume"]=fplt.volume_ocv(data['volume'], ax = axo)
    plots["indexprice"] = fplt.candlestick_ochl(data['indexprice'], ax = ax1)
    fplt.refresh() # refresh autoscaling when all plots complete    

             

if __name__ == "__main__":
    plots = {}
    fplt.y_pad = 0.07  # pad some extra (for control panel)
    fplt.max_zoom_points = 7    
    ax0,ax1= fplt.create_plot("Study", rows=2, init_zoom_periods=300)
    axo = ax0.overlay()    
    ax0.vb.win.keyPressEvent = on_key
    #ax1.vb.win.keyPressEvent = on_key
    
    load_price_history(symbol="TCS", interval="1D",backtest_date= "2023-09-06")
    update_chart()
    fplt.show()


`

**Comment by highfestiva** [comment]
The reason is they contain data which are present in one data set and not the other, which causes an uneven number of candles between the two axes. Remove the redundant data like so:

```python3
pricedf = pricedf.loc[indexdf.index, :]
indexdf = indexdf.loc[pricedf.index, :]
```

**Reply by rkatikam** [reply]
Thanks for the reply.  I have implemented it in a slightly different way and it started working as expected.  I have made a slightly different change where in I have started using the docks from the dockable.py  example. I have to add a hack to get it working  The hack is in the update_plot() function as below:
  ```
 if data[k] is not None:
            plots[k].ax.vb.win._isMouseLeftDrag = True
            plots[k].update_gfx()
            plots[k].ax.vb.win._isMouseLeftDrag = False
    ax0.vb.pan_x(steps=+100)
    ax1.vb.pan_x(steps=+100)
    ax2.vb.pan_x(steps=+100)

```
If I don't use docks and just create multiple axes and plot(as shared here earlier), I just need to use:

```
if data[k] is not None:
            plots[k].update_gfx()
```

Again I am attaching the full code for your reference:

```
import sys
from PyQt6.QtWidgets import QApplication, QMainWindow, QLabel, QWidget
from PyQt6.QtGui import QKeyEvent
from pyqtgraph.dockarea import DockArea, Dock
from PyQt6.QtCore import QDateTime, Qt
import yfinance as yf
import pandas as pd
import finplot as fplt

stock_ohlcv_data = None
index_ohlcv_data = None
backtestdate_Idx = None
backtestdate_Idx1 = None
min_length = None
plots = {}


def load_price_history(symbol, interval, backtest_date):   
    global backtestdate_Idx, stock_ohlcv_data, index_ohlcv_data, backtestdate_Idx1, min_length
    # print("Getting data for: ",symbol)
    # print("Back test Date: ",backtest_date)
    stock_ohlcv_data = yf.download(symbol + ".NS", interval=interval)
    stock_ohlcv_data.dropna()
    stock_ohlcv_data["time"] = pd.to_datetime(stock_ohlcv_data.index)
    stock_ohlcv_data = stock_ohlcv_data.drop(["Adj Close"], axis=1)
    # print(stock_ohlcv_data)
    backtestdate_Idx = stock_ohlcv_data.index.get_loc(backtest_date)
    # print("Starting Index of the Backtes Date:",backtestdate_Idx)
    index_ohlcv_data = yf.download("^NSEI", interval=interval)
    index_ohlcv_data.dropna()
    index_ohlcv_data["time"] = pd.to_datetime(index_ohlcv_data.index)
    index_ohlcv_data = index_ohlcv_data.drop(["Adj Close"], axis=1)
    # print(index_ohlcv_data)
    backtestdate_Idx1 = index_ohlcv_data.index.get_loc(backtest_date)
    # return stock_ohlcv_data
    min_length = min(backtestdate_Idx, backtestdate_Idx1)


def compute_data():
    global backtestdate_Idx, backtestdate_Idx1, min_length
    pricedf = stock_ohlcv_data[backtestdate_Idx - min_length : backtestdate_Idx]
    indexdf = index_ohlcv_data[backtestdate_Idx1 - min_length : backtestdate_Idx1]
    # print(pricedf)
    # print(indexdf)
    price = pricedf["Open Close High Low".split()]
    ma20 = pricedf.Close.rolling(20).mean()
    ma50 = pricedf.Close.rolling(50).mean()
    volume = pricedf["Open Close Volume".split()]
    index = indexdf["Open Close High Low".split()]
    plot_data = {
        "stockprice": price,
        "ma20": ma20,
        "ma50": ma50,
        "volume": volume,
        "indexprice": index,
    }
    return plot_data

def update_plot():
    data = compute_data()
    # print(data)
    for k in data:
        # print(data[k])
        if data[k] is not None:
            plots[k].update_data(data[k], gfx=False)
        if data[k] is not None:
            plots[k].ax.vb.win._isMouseLeftDrag = True
            plots[k].update_gfx()
            plots[k].ax.vb.win._isMouseLeftDrag = False
    ax0.vb.pan_x(steps=+100)
    ax1.vb.pan_x(steps=+100)
    ax2.vb.pan_x(steps=+100)


def update_chart():
    ax0.reset()  # remove previous plots
    ax1.reset()  # remove previous plots
    ax2.reset()  # remove previous plots

    data = compute_data()
    global plots

    plots["indexprice"] = fplt.candlestick_ochl(data["indexprice"], ax=ax0)
    plots["ma20"] = fplt.plot(data["ma20"], legend="MA-20", ax=ax1)
    plots["ma50"] = fplt.plot(data["ma50"], legend="MA-50", ax=ax1)
    plots["stockprice"] = fplt.candlestick_ochl(data["stockprice"], ax=ax1)
    plots["volume"] = fplt.volume_ocv(data["volume"], ax=ax2)
    fplt.refresh()  # refresh autoscaling when all plots complete


class KeyboardEventApp(QMainWindow):

    def __init__(self):
        super().__init__()

        self.setWindowTitle("Keyboard Event Example")
        self.setGeometry(100, 100, 400, 200)
       

    def keyReleaseEvent(self, event):
        global backtestdate_Idx, backtestdate_Idx1
        if isinstance(event, QKeyEvent):
            # print(event.key())
            key = event.key()
            # key_text = event.key()
            if key == Qt.Key.Key_Right:
                backtestdate_Idx += 1
                backtestdate_Idx1 += 1
                print("Right Key. New Index=", backtestdate_Idx)
                update_plot()
            elif key == Qt.Key.Key_Left:
                backtestdate_Idx -= 1
                backtestdate_Idx1 -= 1
                print("Left Key. New Index=", backtestdate_Idx)
                update_plot()


def on_key(event):
    # test for a specific key
    if event.key() == Qt.Key.Key_Right:
        print("Got right key on flpt")
    if event.key() == Qt.Key.Key_Left:
        print("Got left key on flpt")


if __name__ == "__main__":
    app = QApplication(sys.argv)
    win = KeyboardEventApp()
    area = DockArea()
    win.setCentralWidget(area)
    win.resize(1600, 800)
    win.setWindowTitle("Study charts")

    dock_0 = Dock("Index", size=(1000, 100), closable=False)
    dock_1 = Dock("Stock", size=(1000, 100), closable=False)
    dock_2 = Dock("Volume", size=(1000, 100), closable=False)
    area.addDock(dock_0)
    area.addDock(dock_1)
    area.addDock(dock_2)
    ax0, ax1, ax2 = fplt.create_plot_widget(master=area, rows=3, init_zoom_periods=300)
    area.axs = [ax0, ax1, ax2]
    dock_0.addWidget(ax0.ax_widget, 1, 0, 1, 2)
    dock_1.addWidget(ax1.ax_widget, 1, 0, 1, 2)
    dock_2.addWidget(ax2.ax_widget, 1, 0, 1, 2)
    ax1.setXLink(ax0)
    ax2.setXLink(ax0)
    win.axs = [ax0, ax1, ax2]
    dock_0.keyPressEvent = on_key
    dock_1.keyPressEvent = on_key
    dock_0.setFocus()

    load_price_history("SDBL", "1D", "2023-09-06")
    update_chart()

    fplt.show(qt_exec=False)  # prepares plots when they're all setup
    win.show()
    app.exec()


```

**Comment by highfestiva** [comment]
Don't know what you're trying to do, but probably you want to disable the default key handler.

```python3
    def dummy(*args, **kwargs):
        pass
    fplt._key_pressed = dummy
```

Then just using the most basic `update_plot()` does at least something for me.

```python
def update_plot():
    data = compute_data()
    for k in data:
        if data[k] is not None:
            plots[k].update_data(data[k])
```
